src/server/main_service.cc

```diff
@@ -1792,97 +1771,80 @@ void Service::EvalInternal(CmdArgList args, const EvalArgs& eval_args, Interpret
   if (!params)
     return cntx->SendError(facade::kScriptNotFound);
 
-  string error;
-
   DCHECK(!cntx->conn_state.script_info);  // we should not call eval from the script.
 
   // TODO: to determine whether the script is RO by scanning all "redis.p?call" calls
   // and checking whether all invocations consist of RO commands.
   // we can do it once during script insertion into script mgr.
   auto& sinfo = cntx->conn_state.script_info;
   sinfo = make_unique<ConnectionState::ScriptInfo>();
-  sinfo->keys.reserve(eval_args.keys.size());
-
-  optional<ShardId> sid;
-
-  UniqueSlotChecker slot_checker;
-  for (size_t i = 0; i < eval_args.keys.size(); ++i) {
-    string_view key = ArgS(eval_args.keys, i);
-    slot_checker.Add(key);
-    sinfo->keys.insert(KeyLockArgs::GetLockKey(key));
-
-    ShardId cur_sid = Shard(key, shard_count());
-    if (i == 0) {
-      sid = cur_sid;
-    }
-    if (sid.has_value() && *sid != cur_sid) {
-      sid = nullopt;
-    }
-  }
-
   sinfo->async_cmds_heap_limit = absl::GetFlag(FLAGS_multi_eval_squash_buffer);
+
   Transaction* tx = cntx->transaction;
   CHECK(tx != nullptr);
 
   interpreter->SetGlobalArray("KEYS", eval_args.keys);
   interpreter->SetGlobalArray("ARGV", eval_args.args);
+  interpreter->SetRedisFunc(
+      [cntx, this](Interpreter::CallArgs args) { CallFromScript(cntx, args); });
 
   absl::Cleanup clean = [interpreter, &sinfo]() {
     interpreter->ResetStack();
     sinfo.reset();
   };
 
+  string error;
   Interpreter::RunResult result;
 
-  if (CanRunSingleShardMulti(sid, *params, *tx)) {
-    // If script runs on a single shard, we run it remotely to save hops.
-    interpreter->SetRedisFunc([cntx, this](Interpreter::CallArgs args) {
-      // Disable squashing, as we're using the squashing mechanism to run remotely.
-      args.async = false;
-      CallFromScript(cntx, args);
-    });
+  bool embedded = tx->GetMultiMode() != Transaction::NOT_DETERMINED;
```

Comment:
rename embedded to `was_undetermined` and reverse the condition here and below

Reply:
`was_undetermined` doesn't explain why it should be so, whether it's embedded or not is what we care about and we check it by multi mode determinedness